cudax/include/cuda/experimental/__memory_resource/any_resource.cuh

```diff
@@ -41,310 +41,847 @@
+template <class... _Properties>
+struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES async_resource_ref
+    : basic_any<__iasync_resource<_Properties...>&>
+    , __with_try_get_property<async_resource_ref<_Properties...>>
+{
+  using async_resource_ref::basic_any::basic_any;
+
+  // Conversions from the resource_ref types in cuda::mr is not supported.
+  template <class... _OtherProperties>
+  async_resource_ref(_CUDA_VMR::async_resource_ref<_OtherProperties...>) = delete;
+
+private:
+  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
+                "The properties of cuda::experimental::async_resource_ref must contain at least one execution space "
+                "property!");
+  using async_resource_ref::basic_any::interface;
+};
+
+_CCCL_TEMPLATE(class... _Properties, class _Resource)
+_CCCL_REQUIRES(mr::resource_with<_Resource, _Properties...>)
+resource_ref<_Properties...> __as_resource_ref(_Resource& __mr) noexcept
+{
+  return resource_ref<_Properties...>(__mr);
+}
+
+template <class... _Properties>
+resource_ref<_Properties...> __as_resource_ref(resource_ref<_Properties...> __mr) noexcept
+{
+  return __mr;
+}
+
+template <class... _Properties>
+resource_ref<_Properties...> __as_resource_ref(async_resource_ref<_Properties...> __mr) noexcept
+{
+  return __mr;
+}
+
+template <class... _Properties, mr::_AllocType _Alloc_type>
+mr::resource_ref<_Properties...> __as_resource_ref(mr::basic_resource_ref<_Alloc_type, _Properties...> __mr) noexcept
+{
+  return __mr;
+}
+
+#else // ^^^ !_CCCL_DOXYGEN_INVOKED ^^^ / vvv _CCCL_DOXYGEN_INVOKED vvv
+
+enum class _ResourceKind
+{
+  _Synchronous,
+  _Asynchronous
+};
+
+//! @rst
+//! Type erased wrapper around a `resource` or an `async_resource`
+//! --------------------------------------------------------------
+//!
+//! ``basic_any_resource`` wraps any given :ref:`resource
+//! <libcudacxx-extended-api-memory-resources-resource>` that satisfies the
+//! required properties. It owns the contained resource, taking care of
+//! construction / destruction. This makes it especially suited for use in e.g.
+//! container types that need to ensure that the lifetime of the container
+//! exceeds the lifetime of the memory resource used to allocate the storage
+//!
+//! ``basic_any_resource`` models the ``cuda::std::regular`` concept.
+//! @endrst
+//!
+//! @tparam _Kind Either `_ResourceKind::_Synchronous` for `any_resource`, or
+//! `_ResourceKind::_Asynchronous` for `any_async_resource`.
+//! @tparam _Properties A pack of property types that a memory resource must
+//! provide in order to be storable in instances of this `basic_any_resource`
+//! type.
+//!
+//! @sa any_resource
+//! @sa any_async_resource
+//! @sa resource_ref
+//! @sa async_resource_ref
+template <_ResourceKind _Kind, class... _Properties>
+class basic_any_resource
+{
+public:
+  //! @brief Constructs a \c basic_any_resource with no value
+  //! @post `has_value()` is `false`
+  basic_any_resource() noexcept;
```

Comment:
Maybe I am stupid, where is the implementation?

Reply:
this whole block, from line 417 to line 861, is strictly for Doxygen's benefit. the actually implementation of, eg, `any_resource` is the following from line 301:

```c++
template <class... _Properties>
struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES any_resource
    : basic_any<__iresource<_Properties...>>
    , __with_try_get_property<any_resource<_Properties...>>
{
  using any_resource::basic_any::basic_any;

  // any_async_resource is convertible to any_resource
  _CCCL_TEMPLATE(class... _OtherProperties)
  _CCCL_REQUIRES(_CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__type_set<_OtherProperties...>, _Properties...>)
  any_resource(any_async_resource<_OtherProperties...> __other) noexcept
      : basic_any<__iresource<_Properties...>>(static_cast<basic_any<__iasync_resource<_OtherProperties...>>&>(__other))
  {}

private:
  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
                "The properties of cuda::experimental::any_resource must contain at least one execution space "
                "property!");
  using any_resource::basic_any::interface;
};
```

I felt that Doxygen comments for this class template would not be very enlightening, so instead i mocked up the user-facing interface and added doc comments to that.